src/simulations/collect.jl

```diff
@@ -0,0 +1,340 @@
+* `parallel=false` : Only when `replicates>0`. Run replicate simulations in parallel.
+"""
+function run! end
+
+run!(model::ABM, agent_step!, n; kwargs...) =
+run!(model::ABM, agent_step!, dummystep, n; kwargs...) =
+
+function run!(model::ABM, agent_step!, model_step!, n;
+  when=1:n, agent_properties=nothing, model_properties=nothing,
+  aggregation_dict=nothing, replicates::Int=0, parallel::Bool=false
+  )
+
+  if replicates > 0
+    if parallel
+      dataall = parallel_replicates(model, agent_step!, model_step!, n, when;
+      agent_properties=agent_properties, model_properties=model_properties, aggregation_dict=aggregation_dict, replicates=replicates)
+    else
+      dataall = series_replicates(model, agent_step!, model_step!, n, when;
+      agent_properties=agent_properties, model_properties=model_properties, aggregation_dict=aggregation_dict, replicates=replicates)
+    end
+    return dataall
+  else
+    df = run!(model, agent_step!, model_step!, n;
+    when=when, agent_properties=agent_properties,
+    model_properties=model_properties, aggregation_dict=aggregation_dict)
+    return df
+  end
+end
+
+###################################################
+# core data collection functions per step
+###################################################
+"""
+    collect_agent_data(model::ABM, properties::Vector, step = 0) → df
+
+Collect agent properties into a dataframe. `properties` can have symbols (fields) or
+functions that take an agent as input and output a value. `step` is given only
+so that the function adds the correct number in the `step` column.
+"""
+function collect_agent_data(model::ABM, properties::AbstractArray, step::Int = 0)
+  dd = DataFrame()
+  dd[!, :id] = collect(keys(model.agents))
+  for fn in properties
+    colname = typeof(fn) <: Function ? Symbol(fn) : fn
+    dd[!, colname] = get_data.(values(model.agents), fn)
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+collect_agent_data(model::ABM, properties::Nothing, step::Int=0) = DataFrame()
+
+"""
+    collect_model_data(model::ABM, properties::Vector, step = 0) → df
+
+Same as [`collect_agent_data`](@ref) but for model data instead.
+"""
+function collect_model_data(model::ABM, properties::AbstractArray, step::Int = 0)
+  dd = DataFrame()
+  for fn in properties
+    r =  get_data(model, fn)
+    colname = typeof(fn) <: Function ? Symbol(fn) : fn
+    dd[!, colname] = [r]
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+collect_model_data(model::ABM, properties::Nothing, step::Int=0) = DataFrame()
+
+"""
+    aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+
+Aggregate `df` columns  with some function(s) specified in `aggregation_dict`.
+Each key in `aggregation_dict` is a column name (Symbol), and each value is
+an array of function to aggregate that column.
+
+Aggregation occurs per step.
+"""
+function aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
```

Comment:
Scenario: `Agent5`, `agent_properties=nothing`, `aggregation_dict=Dict(:weight => [mean])`.

`collect_agent_data!` is called, but since there is no `agent_properties`, `dft = DataFrame()`.

Thus, `dfnames` here will return `0-element Array{Symbol,1}`, thus  `aggregate_data` will return nothing.

Reply:
An assertion that `aggregation_dict` must be `nothing` when `agent_properties` is nothing should solve such scenario.

Reply:
But what if all I want is `mean(weight)` an an output?

Reply:
Then `agent_properties=[:weight]`, `aggregation_dict=Dict(:weight =>[mean])`

Reply:
then you _must_ give `agent_properties=[:weight]` ?

Reply:
I'll write up an example to test that workflow.

Reply:
Aha. Now I see a good point. `agent_properties` is _reduntant_ if `aggregation_dict` is given **EXCEPT** in the scenario where `agent_properties` contains functions. `aggregation_dict` doesn't know if a column of the dataframe came from applying a function or collecting a field.

Reply:
Let's not mess with all these complexities and expect that `agent_properties` must not be nothing if `aggregation_dict` is not nothing...?

Reply:
A little redundancy here and there is a fair price for a simple source code.

Reply:
That may indeed be the outcome. I'll take that as the assumption and work on some tests surrounding this.